File: packages/shared/src/types/theatre.ts
/** Theatre type */
export type TheatreType = "general" | "orthopaedic" | "cardiac" | "neuro" | "ophthalmic" | "obstetric" | "ent" | "dental" | "minor";

/** Theatre status */
export type TheatreStatus = "available" | "in_use" | "cleaning" | "maintenance" | "closed";

/** Surgical priority */
export type SurgicalPriority = "emergency" | "urgent" | "elective";

/** Anaesthesia type */
export type AnaesthesiaType = "general" | "spinal" | "epidural" | "regional" | "local" | "sedation";

/** Surgical case status */
export type SurgicalCaseStatus = "scheduled" | "preop" | "in_progress" | "in_recovery" | "completed" | "cancelled" | "postponed";

/** Theatre creation */
export interface TheatreCreate {
  name: string;
  theatre_type?: TheatreType;
  floor?: string | null;
  equipment?: Record<string, unknown> | null;
  notes?: string | null;
}

/** Theatre response */
export interface TheatreResponse {
  id: string;
  name: string;
  code: string;
  theatre_type: string;
  status: string;
  floor: string | null;
  equipment: Record<string, unknown> | null;
  is_active: boolean;
  notes: string | null;
}

/** Surgical case creation */
export interface SurgicalCaseCreate {
  patient_id: string;
  encounter_id?: string | null;
  theatre_id?: string | null;
  scheduled_date: string;
  estimated_duration_min?: number;
  priority?: SurgicalPriority;
  procedure_name: string;
  procedure_code?: string | null;
  laterality?: string | null;
  diagnosis?: string | null;
  anaesthesia_type?: AnaesthesiaType | null;
  anaesthetist_id?: string | null;
  lead_surgeon_id: string;
  assistant_surgeon_id?: string | null;
  scrub_nurse_id?: string | null;
  circulating_nurse_id?: string | null;
  notes?: string | null;
}

/** Surgical case response */
export interface SurgicalCaseResponse {
  id: string;
  case_number: string;
  patient_id: string;
  encounter_id: string | null;
  theatre_id: string | null;
  scheduled_date: string;
  estimated_duration_min: number;
  priority: string;
  procedure_name: string;
  procedure_code: string | null;
  laterality: string | null;
  diagnosis: string | null;
  anaesthesia_type: string | null;
  anaesthetist_id: string | null;
  lead_surgeon_id: string;
  assistant_surgeon_id: string | null;
  scrub_nurse_id: string | null;
  circulating_nurse_id: string | null;
  status: string;
  preop_start: string | null;
  surgery_start: string | null;
  surgery_end: string | null;
  recovery_start: string | null;
  recovery_end: string | null;
  operative_findings: string | null;
  operative_notes: string | null;
  specimens: Record<string, unknown> | null;
  implants: Record<string, unknown> | null;
  complications: string | null;
  blood_loss_ml: number | null;
  postop_instructions: string | null;
  preop_checklist: Record<string, unknown> | null;
  timeout_checklist: Record<string, unknown> | null;
  signout_checklist: Record<string, unknown> | null;
  cancellation_reason: string | null;
  notes: string | null;
  created_at: string;
}

/** Surgical case list item */
export interface SurgicalCaseListItem {
  id: string;
  case_number: string;
  patient_id: string;
  patient_name: string | null;
  theatre_name: string | null;
  scheduled_date: string;
  procedure_name: string;
  priority: string;
  lead_surgeon_name: string | null;
  status: string;
}

/** Surgical case list response */
export interface SurgicalCaseListResponse {
  items: SurgicalCaseListItem[];
  total: number;
}

/** Operative notes request */
export interface OperativeNotesRequest {
  operative_findings?: string | null;
  operative_notes?: string | null;
  specimens?: Record<string, unknown> | null;
  implants?: Record<string, unknown> | null;
  complications?: string | null;
  blood_loss_ml?: number | null;
  postop_instructions?: string | null;
}

/** Theatre summary */
export interface TheatreSummary {
  total_theatres: number;
  available_theatres: number;
  in_use_theatres: number;
  scheduled_today: number;
  completed_today: number;
  emergency_cases: number;
  cancelled_today: number;
}

File: services/api-gateway/app/schemas/theatre.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

# ── Theatres ─────────────────────────────────────────────────────────────────


class TheatreCreate(BaseModel):
    """Schema for creating an operating theatre."""

    name: str = Field(..., max_length=100)
    theatre_type: str = Field(
        default="general",
        pattern=r"^(general|orthopaedic|cardiac|neuro|ophthalmic|obstetric|ent|dental|minor)$",
    )
    floor: str | None = Field(None, max_length=20)
    equipment: dict | None = None
    notes: str | None = Field(None, max_length=1000)


class TheatreResponse(BaseModel):
    """Schema for theatre API responses."""

    id: uuid.UUID
    name: str
    code: str
    theatre_type: str
    status: str
    floor: str | None
    equipment: dict | None
    is_active: bool
    notes: str | None

    model_config = {"from_attributes": True}


# ── Surgical Cases ───────────────────────────────────────────────────────────


class SurgicalCaseCreate(BaseModel):
    """Schema for scheduling a surgical case."""

    patient_id: uuid.UUID
    encounter_id: uuid.UUID | None = None
    theatre_id: uuid.UUID | None = None
    scheduled_date: datetime
    estimated_duration_min: int = Field(60, ge=15, le=1440)
    priority: str = Field(
        default="elective",
        pattern=r"^(emergency|urgent|elective)$",
    )
    procedure_name: str = Field(..., max_length=500)
    procedure_code: str | None = Field(None, max_length=20)
    laterality: str | None = Field(None, pattern=r"^(left|right|bilateral|na)$")
    diagnosis: str | None = Field(None, max_length=2000)
    anaesthesia_type: str | None = Field(
        None,
        pattern=r"^(general|spinal|epidural|regional|local|sedation)$",
    )
    anaesthetist_id: uuid.UUID | None = None
    lead_surgeon_id: uuid.UUID
    assistant_surgeon_id: uuid.UUID | None = None
    scrub_nurse_id: uuid.UUID | None = None
    circulating_nurse_id: uuid.UUID | None = None
    notes: str | None = Field(None, max_length=2000)


class SurgicalCaseResponse(BaseModel):
    """Schema for surgical case API responses."""

    id: uuid.UUID
    case_number: str
    patient_id: uuid.UUID
    encounter_id: uuid.UUID | None
    theatre_id: uuid.UUID | None
    scheduled_date: datetime
    estimated_duration_min: int
    priority: str
    procedure_name: str
    procedure_code: str | None
    laterality: str | None
    diagnosis: str | None
    anaesthesia_type: str | None
    anaesthetist_id: uuid.UUID | None
    lead_surgeon_id: uuid.UUID
    assistant_surgeon_id: uuid.UUID | None
    scrub_nurse_id: uuid.UUID | None
    circulating_nurse_id: uuid.UUID | None
    status: str
    preop_start: datetime | None
    surgery_start: datetime | None
    surgery_end: datetime | None
    recovery_start: datetime | None
    recovery_end: datetime | None
    operative_findings: str | None
    operative_notes: str | None
    specimens: dict | None
    implants: dict | None
    complications: str | None
    blood_loss_ml: int | None
    postop_instructions: str | None
    preop_checklist: dict | None
    timeout_checklist: dict | None
    signout_checklist: dict | None
    cancellation_reason: str | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class SurgicalCaseListItem(BaseModel):
    """Surgical case in list view."""

    id: uuid.UUID
    case_number: str
    patient_id: uuid.UUID
    patient_name: str | None = None
    theatre_name: str | None = None
    scheduled_date: datetime
    procedure_name: str
    priority: str
    lead_surgeon_name: str | None = None
    status: str

    model_config = {"from_attributes": True}


class SurgicalCaseListResponse(BaseModel):
    """Surgical case list response."""

    items: list[SurgicalCaseListItem]
    total: int


class OperativeNotesRequest(BaseModel):
    """Schema for recording operative notes post-surgery."""

    operative_findings: str | None = Field(None, max_length=5000)
    operative_notes: str | None = Field(None, max_length=5000)
    specimens: dict | None = None
    implants: dict | None = None
    complications: str | None = Field(None, max_length=2000)
    blood_loss_ml: int | None = Field(None, ge=0)
    postop_instructions: str | None = Field(None, max_length=2000)


class TheatreSummary(BaseModel):
    """Theatre department summary stats."""

    total_theatres: int = 0
    available_theatres: int = 0
    in_use_theatres: int = 0
    scheduled_today: int = 0
    completed_today: int = 0
    emergency_cases: int = 0
    cancelled_today: int = 0
